ext/atomic/index.md

```diff
@@ -0,0 +1,510 @@
+**MUST** contain at least a `type` member.
+
+For example:
+
+```json
+POST /bulk HTTP/1.1
+Host: example.org
+Content-Type: application/vnd.api+json;ext="https://jsonapi.org/ext/atomic"
+Accept: application/vnd.api+json;ext="https://jsonapi.org/ext/atomic"
+
+{
+  "atomic:operations": [{
+    "op": "add",
+    "href": "/blogPosts",
+    "data": {
+      "type": "articles",
+      "attributes": {
+        "title": "JSON API paints my bikeshed!"
+      }
+    }
+  }]
+}
+```
+
+Note that `href` is used in this example to target a resource collection,
+`blogPosts` in this case, that is distinct from the `type` of the resource.
+
+##### Responses
+
+If the server is able to create a resource with a client-generated ID and its
+representation would be identical to the resource in the operation, then the
+server **MAY** return a result with no `data` member or it **MAY** return a
+result that contains the created resource as `data`.
+
+In all other cases in which the server is able to successfully create the
+resource, the server **MUST** return a result that contains the created resource
+as `data`.
+
+For example:
+
+```json
+HTTP/1.1 200 OK
+Content-Type: application/vnd.api+json;ext="https://jsonapi.org/ext/atomic"
+
+{
+  "atomic:results": [{
+    "data": {
+      "links": {
+        "self": "http://example.com/blogPosts/13"
+      },
+      "type": "articles",
+      "id": "13",
+      "attributes": {
+        "title": "JSON API paints my bikeshed!"
+      }
+    }
+  }]
+}
+```
+
+If the server is unable to create the resource, an appropriate error response
+**MUST** be returned and a response document **SHOULD** be returned that
+contains a top-level `errors` as [described above](#processing-errors).
+
+#### Updating Resources
+
+An operation that updates a resource **MUST** target that resource through the
+operation's `ref` member. The operation **MUST** include an `op` code of
+`"update"`.
+
+For example:
+
+```json
+POST /bulk HTTP/1.1
+Host: example.org
+Content-Type: application/vnd.api+json;ext="https://jsonapi.org/ext/atomic"
+Accept: application/vnd.api+json;ext="https://jsonapi.org/ext/atomic"
+
+{
+  "atomic:operations": [{
+    "op": "update",
+    "ref": {
+      "type": "articles",
+      "id": "13"
+    },
+    "data": {
+      "type": "articles",
+      "id": "13",
+      "attributes": {
+        "title": "To TDD or Not"
+      }
+    }
+  }]
+}
+```
+
+##### Responses
+
+If a server accepts an update but also changes the resource in ways other than
+those specified by the request (for example, updating the `updatedAt` attribute
+or a computed `sha`), then the server **MUST** return a result that includes a
+representation of the updated resource as `data`.
+
+If a server accepts an update and doesn’t update any attributes besides those
+provided, the server **MUST** return a result with no `data` or, if all results
+are empty, the server **MAY** respond with `204 No Content` and no document.
```

Comment:
I recognise that this is the same as for updating a resource in the base spec... but I'll take this opportunity to ask: in both contexts, why forbid including the resource representation if no other attributes have changed? It just makes implementation that little bit more difficult... what is the rationale for making it a MUST rather than a SHOULD or MAY?

Reply:
>  It just makes implementation that little bit more difficult... what is the rationale for making it a MUST rather than a SHOULD or MAY?

@tobyzerner Indeed, this requirement is carried over from the base spec for consistency. The original motivation was to simplify client-side implementations so that responses could simply be ignored, rather than parsed and compared. Do you think that server-side implementations become more difficult with this requirement?